src/input_parsers/OnnxParser.cpp

```diff
@@ -0,0 +1,1265 @@
+}
+
+Set<String> OnnxParser::readInputNames()
+{
+    assert(_network.input().size() >= 1);
+
+    Set<String> initializerNames;
+    for (auto &initNode : _network.initializer())
+    {
+        std::cout << "\ninitialiser: " + initNode.name();
+        initializerNames.insert(initNode.name());
+    }
+
+    Set<String> inputNames;
+    for (auto &inputNode : _network.input())
+    {
+        std::cout << "\ninput: " + inputNode.name();
+        inputNames.insert(inputNode.name());
+    }
+
+    return Set<String>::difference(inputNames, initializerNames);
+}
+
+String OnnxParser::readOutputName()
+{
+    if (_network.output().size() > 1)
+    {
+        String message = "Your model has multiple outputs defined\n";
+        message += "Please specify the name of the output you want to consider using the 'outputName' argument\n";
+        message += "Possible options:";
+        for (auto output : _network.output())
+        {
+            message += " " + output.name();
+        }
+        throw MarabouError(MarabouError::ONNX_PARSE_ERROR, message.ascii());
+    }
+    std::cout << "\noutput: " + _network.output()[0].name();
+    return _network.output()[0].name();
+}
+
+void OnnxParser::initialiseShapeMap()
+{
+    // Add shapes for inputs
+    for (auto input : _network.input())
+    {
+        String inputName = input.name();
+        _shapeMap.insert(inputName, shapeOfInput(input));
+
+        // If this is one of the specified inputs, create new variables
+        if (_inputNames.exists(inputName))
+        {
+            _processedNodes.insert(inputName);
+            _numberOfFoundInputs += 1;
+            Vector<Variable> nodeVars = makeNodeVariables(inputName, true);
+        }
+    }
+
+    // Add shapes for constants
+    for (auto constant : _network.initializer())
+    {
+        String constantName = constant.name();
+        _shapeMap.insert(constantName, shapeOfConstant(constant));
+        _processedNodes.insert(constantName);
+    }
+}
+
+void OnnxParser::validateAllInputsAndOutputsFound()
+{
+    // By this point, all input variables need to have been found
+    if (_numberOfFoundInputs != _inputNames.size())
+    {
+        String errorMessage = "These input variables could not be found:";
+        for (String inputName : _inputNames)
+        {
+            if (!_varMap.exists(inputName))
+            {
+                String space = " ";
+                errorMessage += space + inputName;
+            }
+        }
+        throw MarabouError(MarabouError::ONNX_PARSE_ERROR, errorMessage.ascii());
+    }
+
+    // Mark the output variables
+    for (Variable var : _varMap[_outputName])
+    {
+        _outputVars.append(var);
+    }
+}
+
+void OnnxParser::processGraph(Set<String> &inputNames, String &outputName, InputQuery& query)
+{
+    _inputNames = inputNames;
+    _outputName = outputName;
+
+    initialiseShapeMap();
+    processNode(outputName, true);
```

Comment:
Does this assume that there is only one network output?

Reply:
Yes, not a feature currently supported at the moment (was added to the Python version after I started this PR). Added a doc string to this function explaining this.

Reply:
Sounds good.